src/HealthChecks.UI.Client/UIResponseWriter.cs

```diff
@@ -39,10 +40,16 @@ public static Task WriteHealthCheckUIResponse(HttpContext httpContext, HealthRep
                 var uiReport = UIHealthReport
                     .CreateFrom(report);
 
-                response = JsonSerializer.Serialize(uiReport, settings);
+                using var responseStream = new MemoryStream();
+
+                await JsonSerializer.SerializeAsync(responseStream, uiReport, settings);
+                await httpContext.Response.BodyWriter.WriteAsync(responseStream.ToArray());

```

Comment:
Performance degradation due to ToArray() allocation. Do you use any profiler or perf tests?

Reply:
Even with the extra task allocation the memory consumption is less than a half. The ns execution time is a little bit more in the async version but is more than good for me in exchange of not blocking the thread.


// * Summary *

BenchmarkDotNet=v0.12.0, OS=Windows 10.0.18363
Intel Core i7-6700K CPU 4.00GHz (Skylake), 1 CPU, 8 logical and 4 physical cores
.NET Core SDK=3.1.100
  [Host]     : .NET Core 3.1.0 (CoreCLR 4.700.19.56402, CoreFX 4.700.19.56404), X64 RyuJIT
  Job-RSKDBJ : .NET Core 3.1.0 (CoreCLR 4.700.19.56402, CoreFX 4.700.19.56404), X64 RyuJIT

Runtime=.NET Core 3.1

|          Method |     Mean |   Error |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------------- |---------:|--------:|--------:|-------:|------:|------:|----------:|
|  sync_serialize | 184.6 ns | 3.45 ns | 3.39 ns | 0.2792 |     - |     - |    1168 B |
| async_serialize | 273.1 ns | 6.51 ns | 7.23 ns | 0.1278 |     - |     - |     536 B |

Gist: 

https://gist.github.com/CarlosLanderas/2aaeecdaa1f36abe6929ca35d0d339bc

Reply:
Please rerun your benchmarks. Should be
```c#
[GlobalSetup]
public void Setup()
        {
            report = JsonConvert.DeserializeObject<UIHealthReport>(response);
        }
```

Now `report` field is `null`. `Setup` method just never called, you serialize `null` value.

With `[GlobalSetup]`:
``` ini

BenchmarkDotNet=v0.12.0, OS=Windows 10.0.14393.3243 (1607/AnniversaryUpdate/Redstone1)
Intel Core i7-7700 CPU 3.60GHz (Kaby Lake), 1 CPU, 8 logical and 4 physical cores
Frequency=3515625 Hz, Resolution=284.4444 ns, Timer=TSC
.NET Core SDK=3.1.100
  [Host]     : .NET Core 3.1.0 (CoreCLR 4.700.19.56402, CoreFX 4.700.19.56404), X64 RyuJIT  [AttachedDebugger]
  Job-SOELPN : .NET Core 3.1.0 (CoreCLR 4.700.19.56402, CoreFX 4.700.19.56404), X64 RyuJIT

Runtime=.NET Core 3.1  

```
|          Method |     Mean |     Error |    StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------------- |---------:|----------:|----------:|-------:|------:|------:|----------:|
|  sync_serialize | 3.279 us | 0.0460 us | 0.0431 us | 0.7248 |     - |     - |   2.97 KB |
| async_serialize | 7.441 us | 0.1437 us | 0.1344 us | 0.6485 |     - |     - |   2.66 KB |

Reply:
Crap I forgot the setup attribute... :_(, thanks for pointint it out

Well, indeed it still a little bit better. Slightly better allocation values and much better values serializing nulls :D

By the way, the main concern with the UI Client is having async serialization